src/measure_theory/decomposition/signed_hahn.lean

```diff
@@ -0,0 +1,472 @@
+
+To prove this lemma, we will construct a sequence of measurable sets $(A_n)_{n \in \mathbb{N}}$,
+such that, for all $n$, $A_{n + 1} \subseteq i \setminus \bigcup_{k \le n} A_k$ and
+$\frac{1}{a_n + 1} < s(A_{n + 1})$ for some sequence $a_n \to \infty$.
+
+This sequence of sets does not necessarily exist. However, if this sequence terminates; that is,
+there does not exists any sets satisfying the property, the last $A_n$ will be a negative subset
+of negative measure, hence proving our claim.
+
+In the case that the sequence does not terminate, it is easy to see that
+$i \setminus \bigcup_{k = 0}^\infty A_k$ is the required negative set.
+
+To implement this in Lean, we define several auxilary definitions.
+
+- given sets `i`, `j` and the natural number `n`, `exists_one_div_lt s i j n` is the property that
+  there exists a measurable set `k ⊆ i \ j` such that `1 / (n + 1) < s k`.
+- given sets `i`, `j` and that `i \ j` is not negative, `find_exists_one_div_lt s i j` is the
+  least natural number `n` such that `exists_one_div_lt s i j n`. This definition provides the
+  sequence $(a_n)$ in the proof as described above.
+- given sets `i`, `j` and that `i \ j` is not negative, `some_exists_one_div_lt` chooses the set
+  `k` from `exists_one_div_lt s i j (find_exists_one_div_lt s i j)`.
+- lastly, given the set `i`, `restrict_nonpos_seq s i` is the sequence of sets defined inductively
+  where
+    `restrict_nonpos_seq s i 0 = some_exists_one_div_lt s i ∅` and
+    `restrict_nonpos_seq s i (n + 1) = some_exists_one_div_lt s i (⋃ k ≤ n, restrict_nonpos_seq k)`.
+  This definition represents the sequence $(A_n)$ in the proof as described above.
+
+With these definitions, we are able consider the case where the sequence terminates separately,
+allowing us to prove `exists_subset_restrict_nonpos`.
+-/
+
+private def exists_one_div_lt (s : signed_measure α) (i j : set α) (n : ℕ) : Prop :=
+∃ k : set α, k ⊆ i \ j ∧ measurable_set k ∧ (1 / (n + 1) : ℝ) < s k
+
+private lemma exists_nat_one_div_lt_measure_of_not_negative (hi : ¬ s ≤[i \ j] 0) :
+  ∃ (n : ℕ), exists_one_div_lt s i j n :=
+let ⟨k, hj₁, hj₂, hj⟩ := exists_pos_measure_of_not_restrict_le_zero s hi in
+let ⟨n, hn⟩ := exists_nat_one_div_lt hj in ⟨n, k, hj₂, hj₁, hn⟩
+
+private def find_exists_one_div_lt (s : signed_measure α) (i j : set α) : ℕ :=
+if hi : ¬ s ≤[i \ j] 0 then nat.find (exists_nat_one_div_lt_measure_of_not_negative hi) else 0
+
+private lemma find_exists_one_div_lt_spec (hi : ¬ s ≤[i \ j] 0) :
+  exists_one_div_lt s i j (find_exists_one_div_lt s i j) :=
+begin
+  rw [find_exists_one_div_lt, dif_pos hi],
+  convert nat.find_spec _,
+end
+
+private lemma find_exists_one_div_lt_min (hi : ¬ s ≤[i \ j] 0) {m : ℕ}
+  (hm : m < find_exists_one_div_lt s i j) : ¬ exists_one_div_lt s i j m :=
+begin
+  rw [find_exists_one_div_lt, dif_pos hi] at hm,
+  exact nat.find_min _ hm
+end
+
+private def some_exists_one_div_lt (s : signed_measure α) (i j : set α) : set α :=
```

Comment:
Can you add a doc comment to this? I know the linter lets you off, but I think it would be good anyway